scripts/rematch_reads_to_haplotypes.py

```diff
@@ -0,0 +1,318 @@
+            abs(gap_extension_penalty),
+            matrix
+        )
+        return res
+
+def run_alignment(fa_seq, sequence, start_pos, sc_length, hap_cigar, match_score_only = False):
+    """
+    Perform the alignment between two sequences
+    @param fa_seq: The reference sequence
+    @param sequence: The query sequence
+    @param start_pos: The start position of the read
+    @param sc_length: The length of soft clipping
+    @param hap_cigar: The CIGAR string of the haplotype
+    @param match_score_only: If True, return only the match score, for running faster
+    @return: The alignment score, CIGAR string, start position, and query start position
+    """
+
+    if len(fa_seq) == 0 or len(sequence) == 0:
+        return 0, 0, 0
+
+    if match_score_only:
+        # for saving time, get score only
+        return align_parasail_local(fa_seq, sequence, match_score_only).score, 0, 0
+    else:
+        # calculate start and end as well
+        alignment = align_parasail_local(fa_seq, sequence, match_score_only)
+        cigar = alignment.cigar.decode.decode('ascii')
+        parsed_cigar = re.match(r'^(\d+)([IS])', cigar)
+        t_gap = int(parsed_cigar.group(1)) if parsed_cigar else 0
+        t_gap += alignment.cigar.beg_query
+
+        start_pos_adjust, end_pos_adjust = adjust_start_end_positions(start_pos - sc_length, t_gap, len(alignment.traceback.ref), hap_cigar)
+
+        return alignment.score, start_pos_adjust, end_pos_adjust
+
+
+def adjust_start_end_positions(start_pos, t_gap, alignmnet_length, hap_cigar_tuples):
+    """
+        Adjust start and end positions according to sc and the beginning of the alignment
+    """
+    adjusted_start_pos = 0
+
+    if t_gap > 0:
+        adjusted_start_pos = t_gap
+
+    adjusted_end_pos = adjusted_start_pos + alignmnet_length
+
+    return start_pos + adjusted_start_pos, start_pos + adjusted_end_pos
+
+
+def find_best_haplotype(region_haps, read, reference):
+    # in case cigar starts with soft clip, we need to adjust the haplotype start
+    affected_haps = set()
+    sc_length_start = read.cigar[0][1] if read.cigar[0][0] == 4 else 0
+    sc_length_end = read.cigar[-1][1] if read.cigar[-1][0] == 4 else 0
+
+    # find the best alignment
+    best_score = -np.inf
+    best_hap = None
+    best_start_point = None
+    best_end_point = None
+    read_seq = read.query_sequence
+    best_hap_start_position = None
+    best_hap_seq = None
+
+    read_start_position = read.reference_start - sc_length_start
+    read_end_position = read.reference_end + sc_length_end
+    # extract the region of the assembly that the read overlaps
+    for hap in region_haps:
+        # only in case we overlap the breakpoint
+        # read.reference_start is the breakpoint position
+        hap_sc_size_start = hap.cigar[0][1] if hap.cigar[0][0] == 4 else 0
+        hap_sc_size_end = hap.cigar[-1][1] if hap.cigar[-1][0] == 4 else 0
+
+        hap_seq = hap.query_sequence
```

Comment:
Maybe better to use hap.get_forward_sequence()

Reply:
@mayalevy - I don't understand. reverse complement sequence of the reversed haplotype is the original haplotype to which everything is aligned in the assembly. Why is it an issue to do reverse complement to the read (if necessary)